### PR TITLE
Tell shader validation when we're targeting vulkan shaders

### PR DESCRIPTION
**Pull request by jarisiru**
Changed --generate-uniform-bindings parameter to --vulkan, and pass the vulkan flag to shader validator where needed.
This change also re-enables the push constant generation.

### CODE REVIEW

File: common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java [hunk 1]
```diff
@@ -156,18 +156,26 @@ public boolean areImagesOfShaderResultsInteresting(
 
   public boolean areShadersValid(
       File shaderJobFile,
-      boolean throwExceptionOnInvalid)
+      boolean throwExceptionOnInvalid,
+      boolean vulkan)
```
Comment (afd) [inline common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java:160]: vulkan -> isVulkan

Much as I don't really like JavaDoc, I do think this one could do with a JavaDoc comment in order to explain that the `isVulkan` parameter is used to instruct the validator to work in Vulkan-aware mode.

File: generate-and-run-shaders/src/main/java/com/graphicsfuzz/generator/ShaderProducer.java [hunk 2]
```diff
@@ -84,7 +84,7 @@ public void run() {
             ns.get("aggressively_complicate_control_flow"),
             ns.get("replace_float_literals"),
             donorsDir,
-            ns.get("generate_uniform_bindings"),
+            ns.get("vulkan"),
```
Comment (afd) [inline generate-and-run-shaders/src/main/java/com/graphicsfuzz/generator/ShaderProducer.java:87]: Comment - no action needed: despite my request for the boolean parameter to be named `isVulkan` I think that `vulkan` is still the right name for the command line argument.

File: generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java [hunk 3]
```diff
@@ -128,6 +128,7 @@ public static void mainHelper(String[] args) throws ArgumentParserException,
     final boolean writeProbabilities = ns.getBoolean("write_probabilities");
     final boolean keepBadVariants = ns.getBoolean("keep_bad_variants");
     final boolean stopOnFail = ns.getBoolean("stop_on_fail");
+    final boolean vulkan = ns.getBoolean("vulkan");
```
Comment (afd) [inline generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java:131]: I would rename the receiving variable to `isVulkan`.

File: generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java [hunk 4]
```diff
@@ -124,7 +123,7 @@ public static void mainHelper(String[] args) throws ArgumentParserException, IOE
     final float bFloat = ns.getFloat("b");
     final float aFloat = ns.getFloat("a");
     final int maxUniforms = ns.getInt("max_uniforms");
-    final boolean generateUniformBindings = ns.getBoolean("generate_uniform_bindings");
+    final boolean generateUniformBindings = ns.getBoolean("vulkan");
```
Comment (afd) [inline generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java:126]: Rename `generateUniformBindings` to `isVulkan`.

File: util/src/main/java/com/graphicsfuzz/util/ToolHelper.java [hunk 5]
```diff
@@ -25,13 +25,32 @@
 public class ToolHelper {
 
   public static ExecResult runValidatorOnShader(ExecHelper.RedirectType redirectType, File file)
-        throws IOException, InterruptedException {
+      throws IOException, InterruptedException {
+    return runValidatorOnShader(redirectType, file, false);
+  }
+
+  public static ExecResult runValidatorOnShader(ExecHelper.RedirectType redirectType, File file,
+                                                boolean vulkan)
```
Comment (afd) [inline util/src/main/java/com/graphicsfuzz/util/ToolHelper.java:33]: Change param to `isVulkan` and add JavaDoc (as per the comment at the start of the review for a similar method.)

File: common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java (hunk 1 above)
Comment (jarisiru) [reply common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java]: done

File: generate-and-run-shaders/src/main/java/com/graphicsfuzz/generator/ShaderProducer.java (hunk 2 above)
Comment (jarisiru) [reply generate-and-run-shaders/src/main/java/com/graphicsfuzz/generator/ShaderProducer.java:87]: check

File: generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java (hunk 3 above)
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java]: done, also in places not mentioned in this request

File: generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java (hunk 4 above)
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java]: dang, I missed this one. done

File: util/src/main/java/com/graphicsfuzz/util/ToolHelper.java (hunk 5 above)
Comment (jarisiru) [reply util/src/main/java/com/graphicsfuzz/util/ToolHelper.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java [hunk 6]
```diff
@@ -109,9 +109,8 @@ private static Namespace parse(String[] args) throws ArgumentParserException {
         .setDefault(0)
         .type(Integer.class);
 
-    parser.addArgument("--generate-uniform-bindings")
-        .help("Put all uniforms in uniform blocks and generate bindings; required for Vulkan "
-            + "compatibility.")
+    parser.addArgument("--vulkan")
+        .help("Generate shader targeting Vulkan")
```
Comment (paulthomson) [inline generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java:113]: Same as above.

File: generator/src/main/java/com/graphicsfuzz/generator/tool/PrepareReference.java [hunk 7]
```diff
@@ -60,9 +60,8 @@ private static Namespace parse(String[] args) throws ArgumentParserException {
         .help("Use WebGL spec.")
         .action(Arguments.storeTrue());
 
-    parser.addArgument("--generate-uniform-bindings")
-        .help("Put all uniforms in uniform blocks and generate bindings; required for Vulkan "
-            + "compatibility.")
+    parser.addArgument("--vulkan")
+        .help("Generate shader targeting Vulkan")
```
Comment (paulthomson) [inline generator/src/main/java/com/graphicsfuzz/generator/tool/PrepareReference.java:64]: Same as above

File: generator/src/main/java/com/graphicsfuzz/generator/tool/Fragment2Compute.java [hunk 8]
```diff
@@ -102,9 +102,8 @@ private static Namespace parse(String[] args) throws ArgumentParserException {
         .help("Seed (unsigned 64 bit long integer) for random number generator.")
         .type(String.class);
 
-    parser.addArgument("--generate-uniform-bindings")
-        .help("Put all uniforms in uniform blocks and generate bindings; required for Vulkan "
-            + "compatibility.")
+    parser.addArgument("--vulkan")
+        .help("Generate shader targeting Vulkan")
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/Fragment2Compute.java]: done

File: common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java [hunk 9]
```diff
@@ -154,20 +154,35 @@ public boolean areImagesOfShaderResultsInteresting(
 
   }
 
+  /**
+   * Are shaders listed in shader job valid?
+   * @param shaderJobFile A shader job to check.
+   * @param throwExceptionOnInvalid Request exception if the validation fails
+   * @param vulkan Tell the validator that this is a vulkan target
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply common/src/main/java/com/graphicsfuzz/common/util/ShaderJobFileOperations.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java [hunk 10]
```diff
@@ -137,9 +137,8 @@ public static void addGeneratorCommonArguments(ArgumentParser parser) {
         .help("Replace float literals with uniforms.")
         .action(Arguments.storeTrue());
 
-    parser.addArgument("--generate-uniform-bindings")
-        .help("Put all uniforms in uniform blocks and generate bindings; required for Vulkan "
-            + "compatibility.")
+    parser.addArgument("--vulkan")
+        .help("Generate shader targeting Vulkan")
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java [hunk 11]
```diff
@@ -224,7 +224,6 @@ public static StringBuilder generateVariant(ShaderJob shaderJob,
         // Get the list of (unique) uniform names in all shaders which
         // are eglible for being turned into push constants (so samplers, etc
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java [hunk 12]
```diff
@@ -247,9 +246,9 @@ public static StringBuilder generateVariant(ShaderJob shaderJob,
           pushConstant = Optional.of(new ArrayList<String>(allUniforms)
               .get(random.nextInt(allUniforms.size())));
         }
-        // Note that by default we pass Optional.empty() here.
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/Generate.java:250]: changed to arraylist; not 100% certain it works as I expect it to

File: generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java [hunk 13]
```diff
@@ -128,6 +128,7 @@ public static void mainHelper(String[] args) throws ArgumentParserException,
     final boolean writeProbabilities = ns.getBoolean("write_probabilities");
     final boolean keepBadVariants = ns.getBoolean("keep_bad_variants");
     final boolean stopOnFail = ns.getBoolean("stop_on_fail");
+    final boolean isVulkan = ns.getBoolean("vulkan");
```
[the remark this answers is not in the document]
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/GenerateShaderFamily.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java (hunk 6 above)
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/KnownValueShaderGenerator.java]: done

File: generator/src/main/java/com/graphicsfuzz/generator/tool/PrepareReference.java (hunk 7 above)
Comment (jarisiru) [reply generator/src/main/java/com/graphicsfuzz/generator/tool/PrepareReference.java]: done

**Pull request merged**